pkg/engine/plugin_host.go

```diff
@@ -31,10 +33,23 @@ type clientLanguageRuntimeHost struct {
 }
 
 func connectToLanguageRuntime(ctx *plugin.Context, address string) (plugin.Host, error) {
+
+	debugMetadata := map[string]interface{}{
+		"mode":    "client",
+		"address": address,
+		"plugin":  "language",
+	}
+
 	// Dial the language runtime.
 	conn, err := grpc.Dial(address, grpc.WithInsecure(),
-		grpc.WithUnaryInterceptor(rpcutil.OpenTracingClientInterceptor()),
-		grpc.WithStreamInterceptor(rpcutil.OpenTracingStreamClientInterceptor()),
+		grpc.WithChainUnaryInterceptor(
+			rpcutil.OpenTracingClientInterceptor(),
+			interceptors.DebugClientInterceptor(debugMetadata),
+		),
+		grpc.WithChainStreamInterceptor(
+			rpcutil.OpenTracingStreamClientInterceptor(),
+			interceptors.DebugStreamClientInterceptor(debugMetadata),
+		),
```

Comment:
Not sure how I feel about always chaining these even when unused. I'd like the env var to control something in this repo, not as a side effect in the library.

Reply:
I hear you, but there's a technical nasty reason for this!

We really only want to instrument `pulumi` CLI here on the server side, and only instrument the clients in the Pulumi CLI. An ENV var would propagate to all sub-processes and enable tracing in servers created via generic Serve. We don't want those traces, if they do come we need FS-level locking when persisting them to a file, or else to disambiguate the file name by process ID. 

It's not possible to solve this more cleanly and directly without taking some deeper changes around our API, with some complicated dance to avoid breaking changes. This is because the generic Serve* function is used to construct all kinds of servers, including  servers for resource providers. What we'd want to do is change the signature of Serve* so that the caller would configure interceptors. So like `ServeWithOptions*` perhaps? And then switch over the call sites to control this.

It felt annoying to me to change the public SDK signatures to get this in, but I could go ahead with it. WDYT?

Reply:
Oh that makes sense. Yeah, I think we'd likely want a new "ServeWithOptions". That would take a struct with optional interceptors, so we can pass in the output of the rpcutil function?

Reply:
Yes! I like that. I'll work it out, sorry it might take me a bit because I'm juggling other things and the PR is useful to me even when unmerged, but hopefully there's no rush here.